src/ucs/datastruct/array.h

```diff
@@ -0,0 +1,224 @@
+ * @param _value_type   Type of array's values (could be anything)
+ */
+#define UCS_ARRAY_DECLARE_TYPE(_name, _index_type, _value_type) \
+    typedef struct { \
+        _value_type       *buffer; \
+        _index_type       length; \
+        _index_type       capacity; \
+    } ucs_array_t(_name); \
+
+
+/**
+ * Declare the function prototypes of an array
+ *
+ * @param _name          Array name
+ * @param _index_type    Type of array's index
+ * @param _value_type    Type of array's values
+ * @param _storage_class Storage class for array's functions (e.g 'static inline')
+ */
+#define UCS_ARRAY_DECLARE_FUNCS(_name, _index_type, _value_type, _storage_class) \
+    _storage_class void \
+    UCS_ARRAY_IDENTIFIER(_name, _init_dynamic)(ucs_array_t(_name) *array); \
+    \
+    _storage_class void \
+    UCS_ARRAY_IDENTIFIER(_name, _cleanup_dynamic)(ucs_array_t(_name) *array); \
+    \
+    _storage_class ucs_status_t \
+    UCS_ARRAY_IDENTIFIER(_name, _reserve)(ucs_array_t(_name) *array, \
+                                          _index_type min_capacity); \
+    \
+    _storage_class ucs_status_t \
+    UCS_ARRAY_IDENTIFIER(_name, _append)(ucs_array_t(_name) *array, \
+                                         _index_type *index_p)
+
+
+/**
+ * Static initializer to create a fixed-length array with existing static buffer
+ * as backing storage. Such array can track the number of elements and check for
+ * overrun, and does not have to be released.
+ *
+ * @param _buffer    Buffer to use as backing store
+ * @param _capacity  Buffer capacity
+ *
+ * Example:
+ *
+ * @code{.c}
+ * int buffer[20];
+ * ucs_array_t(int_array) array = UCS_ARRAY_FIXED_INITIALIZER(buffer);
+ * @endcode
+ */
+#define UCS_ARRAY_FIXED_INITIALIZER(_buffer, _capacity) \
+    { (_buffer), 0, \
+      ((_capacity) & UCS_ARRAY_CAP_MASK) | UCS_ARRAY_CAP_FLAG_FIXED }
+
+
+/**
+ * Expands to array type definition
+ *
+ * @param _name Array name (as passed to UCS_ARRAY_DECLARE)
+ *
+ * Example:
+ *
+ * @code{.c}
+ * ucs_array_t(int_array) my_array;
+ * @endcode
+ */
+#define ucs_array_t(_name) \
+    UCS_ARRAY_IDENTIFIER(_name, _t)
+
+
+/**
+ * Initialize a dynamic array. Such array can grow automatically to accommodate
+ * for more elements.
+ *
+ * @param _name    Array name
+ * @param _array   Pointer to the array to initialize
+ */
+#define ucs_array_init_dynamic(_name, _array) \
+    UCS_ARRAY_IDENTIFIER(_name, _init_dynamic)(_array)
+
+
+/*
+ * Cleanup a dynamic array.
+ *
+ * @param _name    Array name
+ * @param _array   Array to cleanup
+ */
+#define ucs_array_cleanup_dynamic(_name, _array) \
+    UCS_ARRAY_IDENTIFIER(_name, _cleanup_dynamic)(_array)
+
+
+/*
+ * Grow the array memory buffer to the at least required capacity. This does not
+ * change the array length or existing contents, but the backing buffer can be
+ * relocated to another memory area.
+ *
+ * @param _name    Array name
+ * @param _array   Array to reserve buffer for.
```

Comment:
pls, fix align all parameter's descriptions to include or not include a trailing `.`